services/endpointsrv/checks.go

```diff
@@ -0,0 +1,345 @@
+package endpointsrv
+
+import (
+	"bytes"
+	"context"
+	"database/sql"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"net/url"
+	"time"
+
+	"github.com/rs/zerolog/log"
+	"github.com/unweave/unweave/api/types"
+	"github.com/unweave/unweave/db"
+	"go.jetpack.io/typeid"
+)
+
+type endpointChecker struct {
+	checkID string
+	loaded  bool
+	checks  []checkEndpointStep
+}
+
+func newEndpointChecker(checkID string) (*endpointChecker, error) {
+	if checkID == "" {
+		return nil, fmt.Errorf("check id must be provided")
+	}
+
+	checker := &endpointChecker{
+		checkID: checkID,
+	}
+
+	return checker, nil
+}
+
+func (c *endpointChecker) Run(ctx context.Context) error {
+	if !c.loaded {
+		panic("check steps must be loaded before calling run")
+	}
+
+	go func() {
+		for _, check := range c.checks {
+			check.callEndpoint(ctx)
+			check.assertResponse(ctx)
+
+			if check.err != nil {
+				log.Error().Err(check.err).Msg("check failed")
+			}
+
+			log.Debug().
+				Str("check_id", check.checkID).
+				Str("step_id", check.stepID).
+				Str("input", string(check.input)).
+				Str("response", string(check.endpointResponse)).
+				Str("assertion", check.assertion).
```

Comment:
Could we set these as constants here: https://github.com/unweave/unweave/blob/master/api/types/context.go#L8

Reply:
This is code I didn't change. It just appears here because I moved it out of `service.go`.

Do you usually handle those drive-by improvements as part of this sort of PR or in a follow-up?

Reply:
Yeah I know but might as well fix it as we go along. Reduces the overhead.

Reply:
👍 

I'm going to pull the discussion about context keys offline and merge this for now.

It's unclear to me the connection between context keys and this log keys.